boofuzz/primitives/bytes.py

```diff
@@ -91,136 +97,119 @@ class Bytes(BasePrimitive):
-        @type  fuzzable:   bool
-        @param fuzzable:   (Optional, def=True) Enable/disable fuzzing of this primitive
         @type  max_len:    int
         @param max_len:    (Optional, def=None) Maximum string length
-        @type  name:       str
-        @param name:       (Optional, def=None) Specifying a name gives you direct access to a primitive
         """
 
         super(Bytes, self).__init__()
 
-        assert isinstance(value, bytes)
-        self._original_value = value
-        self._value = self._original_value
         self.size = size
         self.max_len = max_len
         if self.size is not None:
             self.max_len = self.size
         self.padding = padding
-        self._fuzzable = fuzzable
-        self._name = name
-        self.this_library = [self._value * 2, self._value * 10, self._value * 100]
 
-    @property
-    def name(self):
-        return self._name
-
-    def mutate(self):
-        """
-        Mutate the primitive by stepping through the fuzz library extended with the "this" library, return False on
-        completion.
-
-        @rtype:  bool
-        @return: True on success, False otherwise.
-        """
+    def mutations(self):
+        for fuzz_value in self._iterate_fuzz_cases():
+            if callable(fuzz_value):
+                yield compose(self._adjust_mutation_for_size, fuzz_value)
+            yield self._adjust_mutation_for_size(fuzz_value=fuzz_value)
 
-        while True:
-            # if we've ran out of mutations, raise the completion flag.
-            if self._mutant_index == self.num_mutations():
-                self._fuzz_complete = True
-
-            # if fuzzing was disabled or complete, and mutate() is called, ensure the original value is restored.
-            if not self._fuzzable or self._fuzz_complete:
-                self._value = self._original_value
-                return False
-
-            if self._mutant_index < len(self._fuzz_library):
-                # stage 1a: replace with _fuzz_library items
-                alreadyDone = 0
-                self._value = self._fuzz_library[self._mutant_index - alreadyDone]
-            elif self._mutant_index < len(self._fuzz_library) + len(self.this_library):
-                # stage 1b: replace with this_library items
-                alreadyDone = len(self._fuzz_library)
-                self._value = self.this_library[self._mutant_index - alreadyDone]
-            elif self._mutant_index < len(self._fuzz_library) + len(self.this_library) + len(self._magic_debug_values):
-                # stage 1c: replace with _magic_debug_value items
-                alreadyDone = len(self._fuzz_library) + len(self.this_library)
-                self._value = self._magic_debug_values[self._mutant_index - alreadyDone]
+    def _adjust_mutation_for_size(self, fuzz_value):
+        if self.size is not None:
+            if len(fuzz_value) > self.size:
```

Comment:
Getting the following error
```python
Traceback (most recent call last):
  File "/home/dev/Documents/boofuzz_requests/dev.py", line 111, in <module>
    fuzz()
  File "/home/dev/Documents/boofuzz_requests/dev.py", line 107, in fuzz
    sess.fuzz()
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 692, in fuzz
    self._main_fuzz_loop((m for path in self._iterate_protocol() for m in self._iterate_single_node(path)))
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 786, in _main_fuzz_loop
    for mutation_context in fuzz_case_iterator:
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 692, in <genexpr>
    self._main_fuzz_loop((m for path in self._iterate_protocol() for m in self._iterate_single_node(path)))
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/sessions.py", line 1367, in _iterate_single_node
    for mutation in self.fuzz_node.mutations():
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/fuzzable_block.py", line 21, in mutations
    for mutation in item.mutations():
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/fuzzable_wrapper.py", line 114, in mutations
    for value in itertools.chain(self._fuzz_object.mutations(), self._fuzz_values):
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/fuzzable_block.py", line 21, in mutations
    for mutation in item.mutations():
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/fuzzable_wrapper.py", line 114, in mutations
    for value in itertools.chain(self._fuzz_object.mutations(), self._fuzz_values):
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/primitives/bytes.py", line 130, in mutations
    yield self._adjust_mutation_for_size(fuzz_value=fuzz_value)
  File "/home/dev/PycharmProjects/boofuzz/boofuzz/primitives/bytes.py", line 134, in _adjust_mutation_for_size
    if len(fuzz_value) > self.size:
TypeError: object of type 'functools.partial' has no len()
```

Reply:
Thanks; I'll start here with unit tests to chase out this error. I did a functional test but probably missed something.

For some discussion, the Bytes primitive is maybe the only one that returns a function from mutations() sometimes. It's the only one with code that modifies the default value. And with the addition of dynamic default values (taking a value from earlier in the protocol and putting it into a field), the default isn't know at the time mutations() is called. Therefore this design decision to have mutations() return a callable. It's powerful, but makes for a confusing primitive implementation.

Reply:
It's a little confusing indeed, so thanks for clearing it up!

Reply:
I'm curious, does this PR fix #400?

Reply:
Actually maybe now that I look at it. That might come up while writing unit tests. However, the "default value" needs to remain constant. So if your "default value" is dynamic and the default data can have different lengths, the ability to calculate the numbers goes out the window.

Reply:
@SR4ven I have an idea to simplify this, hopefully it will get simpler in a bit.

Reply:
@SR4ven The updated version is pushed. Description of changes here.

As you've gathered, `mutations()` can yield a callable that mutates the "normal" value, which may be dynamic.

The previous approach was doing some weird stuff with closures inside the mutations() generator:
```python
        for fuzz_bytes in self._fuzz_strings_1byte:
            i = 0
            keep_going = True
            while keep_going:

                def f(value):
                    nonlocal keep_going
                    if i < len(value):
                        return value[:i] + fuzz_bytes + value[i + 1 :]
                    else:
                        keep_going = False

                yield f
                i += 1
```
The idea was that we wanted to mutate the "value", which can change between protocol runs. BUT, the number of mutations depends on the length of the value. So I did this weird thing with the closure `f` and `nonlocal keepgoing` (nonlocal means the closure can write to the value and the parent function will see it. This still left me with two problems: 1) At the end of the function, we didn't have a value to return. 2) The usage gets weird, as the behavior of the generator depends on the function it yields being called before the next generator iteration.

The compromise I settled on was to have mutations() accept a "default value" (this is how I had it in one of the previous iterators before the PR). So now there's a distinction between "default value" and "normal value". The "default value" must be provided in the protocol definition -- this is either the value as normally provided, or the "default_value" of a `TestCaseSessionReference` object. The "normal value" is usually the same as the default value, but if `TestCaseSessionReference` is used, the normal value is the value referenced by `TestCaseSessionReference`.

Now the `Bytes.mutations()` function uses the default value to control its iteration loops, but it still yields a function that mutates the "normal value".

Reply:
Just pushed finished `unit_tests/tests_bytes.py` which confirms that num_mutations is working for Bytes.